Board/Board.cpp

```diff
@@ -0,0 +1,128 @@
+//
+// Created by Yam on 11/06/2023.
+//
+
+#define DETAILS_DELIMITER ','
+
+#include <fstream>
+#include <sstream>
+#include <typeindex>
+#include <map>
+
+#include "Board.h"
+#include "Piece/Pieces/Pawn/Pawn.h"
+#include "Piece/Pieces/Rook/Rook.h"
+#include "Piece/Pieces/Knight/Knight.h"
+#include "Piece/Pieces/Bishop/Bishop.h"
+#include "Piece/Pieces/Queen/Queen.h"
+#include "Piece/Pieces/King/King.h"
+
+
+Board::Board()
+{
+    Color current_position_color = kWhite;
+    for (int i = 0; i < BOARD_WIDTH; ++i)
+    {
+        for (int j = 0; j < BOARD_LENGTH; ++j)
+        {
+            board[i][j].initializePosition(i, j, current_position_color, nullptr);
+            // Change next position color.
+            current_position_color = (current_position_color) ? kWhite : kBlack;
+        }
+        // Once moving to new row change color again.
+        current_position_color = (current_position_color) ? kWhite : kBlack;
+    }
+}
+
+Board::Board(std::string board_file_path) : Board() {
+    std::ifstream board_file;
+    std::string line;
+
+    board_file.open(board_file_path);
+    if (board_file.is_open() == false)
+    {
+        throw invalidRequest("Unable to open board file.");
+    }
+
+    try {
+        while (std::getline(board_file, line)) {
+            std::string piece_type_string, piece_color_string, x_position_string, y_position_string;
+            std::stringstream piece_info(line);
+
+            // Parse line: piece type, color, x position, y position. e.g. P,B,1,1
+            std::getline(piece_info, piece_type_string, DETAILS_DELIMITER);
+            std::getline(piece_info, piece_color_string, DETAILS_DELIMITER);
+            std::getline(piece_info, x_position_string, DETAILS_DELIMITER);
+            std::getline(piece_info, y_position_string, DETAILS_DELIMITER);
+
+            Color piece_color;
+            switch (piece_color_string.c_str()[0])
+            {
+                case 'W': piece_color = kWhite; break;
+                case 'B': piece_color = kBlack; break;
+                default: throw invalidRequest("Invalid color specifier in board file: " + piece_color_string);
```

Comment:
If you throw exception here the board_file will not close. change it to use RAII class

Reply:
This class is already RAII